docs/develop-resources/api-conventions/validations.md

```diff
@@ -0,0 +1,158 @@
+* This validation shall take the responsibility to guardrail config issues that otherwise could cause ambiguous or unexpected ConfigConnector behavior.
+* We shall build and publish a ConfigConnector error code table to future explain the issue and give a triage guide.
+
+
+## 3. GCP Service Validation
+
+
+* This throws GCP defined errors in runtime. ConfigConnector shall propagate the GCP response errors to the ConfigConnector object’s `status` field. 
+* This validation shall take the responsibility to guide users on fixing the issues due to GCP service requirements, like `SpannerInstance` has `spec.processingUnit` whose value can only increase but not decrease. 
+
+
+# CRD validation
+
+
+## Rule 1: Required/Optional
+
+Required field should use tag `+required`. 
+
+Optional field does not need the tag.
+
+
+```
+type CloudBuildWorkerPoolSpec struct {
+  // +required   
+  PrivatePoolConfig *PrivatePoolV1Config `json:"privatePoolV1Config,omitempty"`
+}
+```
+
+
+## Rule 2: Immutable field
+
+Immutable field should be validated via CEL rule using kubebuilder tag 
+
+
+```
+type PrivatePoolV1Config_NetworkConfigSpec struct {
+        // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="the field is immutable"
+        // Immutable. The network definition that the workers are peered
+        //  to. If this section is left empty, the workers will be peered to
+        //  `WorkerPool.project_id` on the service producer network.
+        PeeredNetworkRef refv1beta1.ComputeNetworkRef `json:"peeredNetworkRef,omitempty"`
+}
+```
+
+Note: 
+
+
+* Kubernetes supports CEL since 1.26 which is the oldest GKE cluster version (written at ). Using CEL shall not cause GKE cluster version issues, but we shall reevaluate this when the condition changes.
+* Some ConfigConnector resources use webhook to validate immutable fields. We can continue using that, but for external contributors CEL is much easier for self-learning. 
+
+We only apply CEL on immutable fields.** More complicated CEL validation requires further discussion.**
+
+One future improvement can be having generate-types detect the “Immutable.” comment and add the kubebuilder tag automatically. 
+
+
+## Rule 3: Parent
+
+Parent field should be required and immutable.
+
+
+### Required
+
+Suggest using the following `inline` struct
+
+
+```
+type CloudBuildWorkerPoolSpec struct {
+  // +required   
+  Parent `json:",inline"`
+}
+
+type Parent struct {
+    // +required      
+    ProjectRef *refv1beta1.ProjectRef `json:"projectRef"`
+
+    // +required
+    Location string `json:"location"`
+}
+```
+
+
+### Immutable
+
+Since most parent is a reference to another resource which can either be `&lt;kind>Ref.external `or `&lt;kind>Ref.name` , the immutable validation needs to be done at the controller validation level. For example, switching `&lt;kind>Ref.external `and `&lt;kind>Ref.name `are allowed if referring to the same resource.
+
+See  for general reference guide;
```

Comment:
Missing link ..?

Reply:
Yeah, because the reference doc is not merged. added a marker as reminder